### ISSUE TITLE
Deletions do not commit to index until a subsequent re-index of another node.

### ISSUE THREAD

**Issue by DccTimbo** [Shazwazza/Examine#55]
I'm using Examine on an MVC web application and have had issues with deletions to the index not commiting to the Lucene index. My project has been setup with a DataService that implements an ISimpleDataService and quite happily indexes the data in my Entity Framework database.

However I have discovered that when issuing a delete operation through the ExamineManager class, the node passed to it is not deleted immediately from the index. The node will eventually delete when a new or existing node is indexed, however until that time the node remains in the search index and appears in search results, which causes a 404 within my application as the corresponding database has already been removed while the index entry lingers in the Lucene index.

I decided to have a peak into the source files for Examine to try and understand what is going on and in the file LuceneIndexer in the project path Examine/LuceneEngine/Providers/LuceneIndexer.cs I believe I have found a bug/erroneous function call which I believe is causing this issue.

On line 1521 of the aforementioned file is the following method.

```
[SecuritySafeCritical]
private void ProcessQueueItem(IndexOperation item, ICollection<IndexedNode> indexedNodes, IndexWriter writer)
{
    switch (item.Operation)
    {
        case IndexOperationType.Add:
            if (ValidateDocument(item.Item.DataToIndex))
            {
                //var added = ProcessIndexQueueItem(item, inMemoryWriter);
                var added = ProcessIndexQueueItem(item, writer);
                indexedNodes.Add(added);
            }
            else
            {
                 //do the delete but no commit - it may or may not exist in the index but since it is not 
                 // valid it should definitely not be there.
                 ProcessDeleteQueueItem(item, writer, false);

                 OnIgnoringNode(new IndexingNodeDataEventArgs(item.Item.DataToIndex, int.Parse(item.Item.Id), null, item.Item.IndexType));
            }
        break;
    case IndexOperationType.Delete:
        ProcessDeleteQueueItem(item, writer, false);
        break;
    default:
        throw new ArgumentOutOfRangeException();
    }
}
```

For the IndexOperationType.Delete case, I believe the final parameter of the ProcessDeleteQueueItem should be set to true, as it is a flag as to whether to commit the change or not.

As it stands I believe that the current action is not being committed until a subsequent re-index action is processed and commits any outstanding actions to the index because the delete operation is not committing its own change to the index.

I've not managed to test this out yet, but was wondering if you could confirm my suspicions or not.

Kind regards,
Tim

**Comment by Shazwazza** [comment]
Hi! Thanks a ton for the research, much appreciated. We don't want each delete to force a commit since that will cause some performance issues, but here is the line that will perform the commit for deletes once the queue is completed: https://github.com/Shazwazza/Examine/blob/dev/Projects/Examine/LuceneEngine/Providers/LuceneIndexer.cs#L1545

Are you running on the latest Examine version? https://www.nuget.org/packages/Examine/0.1.69

**Comment by DccTimbo** [comment]
Hi.

Thanks for the speedy response, and I can confirm I am running the latest version of Examine. Having looked at the code, and in particular the ForceProcessQueueItems method you have highlighted I noticed you raise an event for the NodesIndexed event.

I therefore added events to the following Indexer events.
IndexingError, IndexOperationComplete, IndexDeleted and NodesIndexed.

When I add a node to the index, IndexOperationComplete and NodesIndexed are both raised. When I delete a node, IndexDeleted and IndexOperationComplete fire, but NodesIndexed does not, this suggests to me that the ForceProcessQueueItems method is not firing on my Delete request and may explain why my deletions don't seem to commit. Furthermore the IndexingError event does not fire, so I'm well and truly stumped on this one.

As mentioned in my previous post I am raising a delete request through the ExamineManager class, here is my method I use to do this.

```
public static double DeleteFromIndex(String indexerName, String nodeId)
{
    var timer = new Stopwatch();
    timer.Start();
    if (ExamineManager.Instance.IndexProviderCollection[indexerName] != null)
    {
        if (ExamineManager.Instance.IndexProviderCollection[indexerName].IndexExists())
        {
            var indexer = ExamineManager.Instance.IndexProviderCollection[indexerName] as SimpleDataIndexer;
            indexer.DeleteFromIndex(nodeId);
        }
        else
        {
            ExamineManager.Instance.IndexProviderCollection[indexerName].RebuildIndex();
        }
    }
    timer.Stop();

    return timer.Elapsed.TotalSeconds;
}
```

I appreciate you are a busy man, but if you could offer any suggestions or pointers as to what I may be doing wrong I would greatly appreciate it.

Kind regards,
Tim

**Comment by DccTimbo** [comment]
Hi again,

Since my last comment I have downloaded the 0.1.69 build as a zip file onto my workstation and have discovered that the version of LuceneIndexer.cs differs from that shown on GitHub in the current dev branch, so believe that 0.1.69 has indeed got a bug with regard to deletions.

For example one such difference is the method you highlighted where the commits take place differs in the downloaded / distributed version as it is inside the following conditional if statement.

```
if (indexedNodes.Count > 0)
{
    //commit the changes (this will process the deletes)
    writer.Commit();

    //this is required to ensure the index is written to during the same thread execution
    // if we are in blocking mode, the do the wait
    if (!RunAsync || block)
    {
        writer.WaitForMerges();
    }

    //raise the completed event
    OnNodesIndexed(new IndexedNodesEventArgs(IndexerData, indexedNodes));
}
```

Having added a delete method to the sample web project that is bundled with the project, I discovered that when only Delete commands are issued to the index, they are not committed until a subsequent add / re-index operation is submitted much like in my own project. I therefore decided to step through the source code to trace what was happening.

I discovered that because the if statement listed above is checking a count of completed index operations and processing a delete operation does not add an item to the indexedNodes collection, as per method here:

https://github.com/Shazwazza/Examine/blob/v0.1.69/Projects/Examine/LuceneEngine/Providers/LuceneIndexer.cs#L1522

, the count will remain at 0 when only delete actions are in the queue and so they are not committed to the index at all until an add / re-index operation is issued. 

Furthermore if the application terminates before this happens, the operation is never committed and a rebuild is required the next time the application runs to re-sync the index with the data source.

Removing the if statement resolves the issue, so if this is ready for release it would fix this bug in the 0.1.69 branch.

Kind regards,
Tim

**Comment by Shazwazza** [comment]
Hi, thanks for the investigation, very much appreciated!

Ouch, i can see why this no longer works, based on this rev change here: 1c497dcf07cb3a089afaed45fffb69619d675ff9 

This is unfortunate indeed. I'll have a look to see what can be done and try to get a release out this/next week. Thanks again
